src/oidcc_auth_util.erl

```diff
@@ -0,0 +1,255 @@
+) ->
+    PreferredAuthMethods = maps:get(preferred_auth_methods, Opts, [
+        private_key_jwt,
+        client_secret_jwt,
+        client_secret_post,
+        client_secret_basic,
+        none
+    ]),
+    case select_preferred_auth(PreferredAuthMethods, SupportedAuthMethods) of
+        {ok, AuthMethod} ->
+            case
+                add_authentication(QueryList0, Header0, AuthMethod, AllowAlgorithms, ClientContext)
+            of
+                {ok, {QueryList, Header}} ->
+                    {ok, {QueryList, Header}};
+                {error, _} ->
+                    add_client_authentication(
+                        QueryList0,
+                        Header0,
+                        SupportedAuthMethods -- [atom_to_binary(AuthMethod)],
+                        AllowAlgorithms,
+                        Opts,
+                        ClientContext
+                    )
+            end;
+        {error, Reason} ->
+            {error, Reason}
+    end.
+
+-spec add_authentication(
+    QueryList,
+    Header,
+    AuthMethod,
+    AllowAlgorithms,
+    ClientContext
+) ->
+    {ok, {oidcc_http_util:query_params(), [oidcc_http_util:http_header()]}}
+    | {error, auth_method_not_possible}
+when
+    QueryList :: oidcc_http_util:query_params(),
+    Header :: [oidcc_http_util:http_header()],
+    AuthMethod :: auth_method(),
+    AllowAlgorithms :: [binary()] | undefined,
+    ClientContext :: oidcc_client_context:t().
+add_authentication(
+    QsBodyList,
+    Header,
+    none,
+    _AllowArgs,
+    #oidcc_client_context{client_id = ClientId}
+) ->
+    NewBodyList = [{<<"client_id">>, ClientId} | QsBodyList],
+    {ok, {NewBodyList, Header}};
+add_authentication(
+    _QsBodyList,
+    _Header,
+    _Method,
+    _AllowAlgs,
+    #oidcc_client_context{client_secret = unauthenticated}
+) ->
+    {error, auth_method_not_possible};
+add_authentication(
+    QsBodyList,
+    Header,
+    client_secret_basic,
+    _AllowAlgs,
+    #oidcc_client_context{client_id = ClientId, client_secret = ClientSecret}
+) ->
+    NewHeader = [oidcc_http_util:basic_auth_header(ClientId, ClientSecret) | Header],
+    {ok, {QsBodyList, NewHeader}};
+add_authentication(
+    QsBodyList,
+    Header,
+    client_secret_post,
+    _AllowAlgs,
+    #oidcc_client_context{client_id = ClientId, client_secret = ClientSecret}
+) ->
+    NewBodyList =
+        [{<<"client_id">>, ClientId}, {<<"client_secret">>, ClientSecret} | QsBodyList],
+    {ok, {NewBodyList, Header}};
+add_authentication(
+    QsBodyList,
+    Header,
+    client_secret_jwt,
+    AlgValuesSupported,
+    ClientContext
+) when AlgValuesSupported == []; AlgValuesSupported == undefined ->
+    %% https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata
+    %% Servers SHOULD support RS256.
+    add_authentication(
+        QsBodyList, Header, client_secret_jwt, [<<"RS256">>, <<"HS256">>], ClientContext
```

Comment:
@maennchen I did have one question about this. My read of the spec was that `RS256` should be allowed but not `HS256`, but there's a specific test case which relies on `HS256` being allowed. This was introduced to fix https://github.com/erlef/oidcc/issues/287 but I don't totally understand the original issue.

Reply:
@paulswartz I have to read up on `HS` vs `RS` algorithms. I'm not entirely sure anymore why i defined it this way.

Reply:
@paulswartz Looking into it I'm not sure if this is correct the other way around.

The idea of `client_secret_jwt` is to infer the signature from the `client_secret`. This therefore operates on a shared secret & a MAC. With `RS256` we need a public & private key.

Therefore I'm questioning if `RS256` should be here and not `HS256`.

What part of the specification are you referring to?

Reply:
Let's discuss this separately in #301